flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/consumer/UnionInputGate.java

```diff
@@ -210,12 +211,16 @@ public void requestPartitions() throws IOException, InterruptedException {
 				}
 
 				Iterator<InputGate> inputGateIterator = inputGatesWithData.iterator();
-				final InputGate inputGate = inputGateIterator.next();
+				inputGate = inputGateIterator.next();
 				inputGateIterator.remove();
+			}
 
-				// In case of inputGatesWithData being inaccurate do not block on an empty inputGate, but just poll the data.
-				Optional<BufferOrEvent> bufferOrEvent = inputGate.pollNextBufferOrEvent();
+			// In case of inputGatesWithData being inaccurate do not block on an empty inputGate, but just poll the data.
+			// Do not poll the gate under inputGatesWithData lock, since this can trigger notifications
+			// that could deadlock because of wrong locks taking order.
+			Optional<BufferOrEvent> bufferOrEvent = inputGate.pollNextBufferOrEvent();
 
+			synchronized (inputGatesWithData) {
```

Comment:
I can see that this can fix the deadlock, but it also splits up a previously atomic operation into two parts. I could not spot a concrete problem with it, just wanted to make sure that we don't miss a potential problem here.

Reply:
This was the old behaviour. I have concatenated those two critical sections together in a hotfix/refactoring commit 😳 

There should be no issue with this, since there are only two actors locking on {{inputGatesWithData}}: one reader of the input gate, and the guy/guys notifying that new data has arrived.